// mongoose/lib/schema/index.js
'use strict';

var _commonjsHelpers = require('../../../../_virtual/_commonjsHelpers.js');
var string = require('./string.js');
var number = require('./number.js');
var boolean = require('./boolean.js');
var documentarray = require('./documentarray.js');
var SingleNestedPath = require('./SingleNestedPath.js');
var array = require('./array.js');
var buffer = require('./buffer.js');
var date = require('./date.js');
var objectid = require('./objectid.js');
var mixed = require('./mixed.js');
var decimal128 = require('./decimal128.js');
var map = require('./map.js');

var schema = _commonjsHelpers.createCommonjsModule(function (module, exports) {

exports.String = string;

exports.Number = number;

exports.Boolean = boolean;

exports.DocumentArray = documentarray;

exports.Embedded = SingleNestedPath;

exports.Array = array;

exports.Buffer = buffer;

exports.Date = date;

exports.ObjectId = objectid;

exports.Mixed = mixed;

exports.Decimal128 = exports.Decimal = decimal128;

exports.Map = map;

// alias

exports.Oid = exports.ObjectId;
exports.Object = exports.Mixed;
exports.Bool = exports.Boolean;
exports.ObjectID = exports.ObjectId;
});

module.exports = schema;
