// mongoose/lib/schema/index.mjs
import { createCommonjsModule } from '../../../../_virtual/_commonjsHelpers.js';
import string from './string.mjs';
import number from './number.mjs';
import boolean from './boolean.mjs';
import documentarray from './documentarray.mjs';
import SingleNestedPath_1 from './SingleNestedPath.mjs';
import array from './array.mjs';
import buffer from './buffer.mjs';
import date from './date.mjs';
import objectid from './objectid.mjs';
import mixed from './mixed.mjs';
import decimal128 from './decimal128.mjs';
import map from './map.mjs';

var schema = createCommonjsModule(function (module, exports) {

exports.String = string;

exports.Number = number;

exports.Boolean = boolean;

exports.DocumentArray = documentarray;

exports.Embedded = SingleNestedPath_1;

exports.Array = array;

exports.Buffer = buffer;

exports.Date = date;

exports.ObjectId = objectid;

exports.Mixed = mixed;

exports.Decimal128 = exports.Decimal = decimal128;

exports.Map = map;

// alias

exports.Oid = exports.ObjectId;
exports.Object = exports.Mixed;
exports.Bool = exports.Boolean;
exports.ObjectID = exports.ObjectId;
});

export default schema;
