// @colyseus/core/build/packages/core/src/Server.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var Debug = require('./Debug.js');
var MatchMaker = require('./MatchMaker.js');
var controller = require('./matchmaker/controller.js');
var Room = require('./Room.js');
var Utils = require('./Utils.js');
require('@gamestdio/timer');
var index$1 = require('./discovery/index.js');
var LocalPresence = require('./presence/LocalPresence.js');
var index = require('./matchmaker/driver/index.js');
var Logger = require('./Logger.js');
require('./Protocol.js');
require('events');
require('./Transport.js');
require('@colyseus/schema');
require('./rooms/RelayRoom.js');

class Server {
    transport;
    presence;
    driver;
    processId = Utils.generateId();
    port;
    constructor(options = {}) {
        const { gracefullyShutdown = true } = options;
        this.presence = options.presence || new LocalPresence.LocalPresence();
        this.driver = options.driver || new index.LocalDriver();
        // "presence" option is not used from now on
        delete options.presence;
        this.attach(options);
        // setup matchmaker
        MatchMaker.setup(this.presence, this.driver, this.processId, options.publicAddress);
        if (gracefullyShutdown) {
            Utils.registerGracefulShutdown((err) => this.gracefullyShutdown(true, err));
        }
        if (options.logger) {
            Logger.setLogger(options.logger);
        }
    }
    attach(options) {
        /**
         * Display deprecation warnings for moved Transport options.
         * TODO: Remove me on 0.15
         */
        if (options.pingInterval !== undefined ||
            options.pingMaxRetries !== undefined ||
            options.server !== undefined ||
            options.verifyClient !== undefined) {
            Logger.logger.warn("DEPRECATION WARNING: 'pingInterval', 'pingMaxRetries', 'server', and 'verifyClient' Server options will be permanently moved to WebSocketTransport on v0.15");
            Logger.logger.warn(`new Server({
  transport: new WebSocketTransport({
    pingInterval: ...,
    pingMaxRetries: ...,
    server: ...,
    verifyClient: ...
  })
})`);
            Logger.logger.warn("ðŸ‘‰ Documentation: https://docs.colyseus.io/server/transport/");
        }
        const transport = options.transport || this.getDefaultTransport(options);
        delete options.transport;
        this.transport = transport;
        if (this.transport.server) {
            this.transport.server.once('listening', () => this.registerProcessForDiscovery());
            this.attachMatchMakingRoutes(this.transport.server);
        }
    }
    /**
     * Bind the server into the port specified.
     *
     * @param port
     * @param hostname
     * @param backlog
     * @param listeningListener
     */
    async listen(port, hostname, backlog, listeningListener) {
        this.port = port;
        return new Promise((resolve, reject) => {
            this.transport.server?.on('error', (err) => reject(err));
            this.transport.listen(port, hostname, backlog, (err) => {
                if (listeningListener) {
                    listeningListener(err);
                }
                if (err) {
                    reject(err);
                }
                else {
                    resolve();
                }
            });
        });
    }
    async registerProcessForDiscovery() {
        // register node for proxy/service discovery
        index$1.registerNode(this.presence, {
            port: this.port,
            processId: this.processId,
        });
    }
    /**
     * Define a new type of room for matchmaking.
     *
     * @param name public room identifier for match-making.
     * @param handler Room class definition
     * @param defaultOptions default options for `onCreate`
     */
    define(name, handler, defaultOptions) {
        return MatchMaker.defineRoomType(name, handler, defaultOptions);
    }
    async gracefullyShutdown(exit = true, err) {
        await index$1.unregisterNode(this.presence, {
            port: this.port,
            processId: this.processId,
        });
        try {
            await MatchMaker.gracefullyShutdown();
            this.transport.shutdown();
            this.presence.shutdown();
            this.driver.shutdown();
            await this.onShutdownCallback();
        }
        catch (e) {
            Debug.debugAndPrintError(`error during shutdown: ${e}`);
        }
        finally {
            if (exit) {
                process.exit(err ? 1 : 0);
            }
        }
    }
    /**
     * Add simulated latency between client and server.
     * @param milliseconds round trip latency in milliseconds.
     */
    simulateLatency(milliseconds) {
        Logger.logger.warn(`ðŸ“¶ï¸â— Colyseus latency simulation enabled â†’ ${milliseconds}ms latency for round trip.`);
        const halfwayMS = (milliseconds / 2);
        this.transport.simulateLatency(halfwayMS);
        /* tslint:disable:no-string-literal */
        const _onMessage = Room.Room.prototype['_onMessage'];
        /* tslint:disable:no-string-literal */
        Room.Room.prototype['_onMessage'] = function (client, buffer) {
            // uWebSockets.js: duplicate buffer because it is cleared at native layer before the timeout.
            const cachedBuffer = Buffer.from(buffer);
            setTimeout(() => _onMessage.call(this, client, cachedBuffer), halfwayMS);
        };
    }
    /**
     * Register a callback that is going to be executed before the server shuts down.
     * @param callback
     */
    onShutdown(callback) {
        this.onShutdownCallback = callback;
    }
    getDefaultTransport(_) {
        throw new Error("Please provide a 'transport' layer. Default transport not set.");
    }
    onShutdownCallback = () => Promise.resolve();
    attachMatchMakingRoutes(server) {
        const listeners = server.listeners('request').slice(0);
        server.removeAllListeners('request');
        server.on('request', (req, res) => {
            if (req.url.indexOf(`/${controller.matchmakeRoute}`) !== -1) {
                Debug.debugMatchMaking('received matchmake request: %s', req.url);
                this.handleMatchMakeRequest(req, res);
            }
            else {
                for (let i = 0, l = listeners.length; i < l; i++) {
                    listeners[i].call(server, req, res);
                }
            }
        });
    }
    async handleMatchMakeRequest(req, res) {
        const headers = {
            'Access-Control-Allow-Headers': 'Origin, X-Requested-With, Content-Type, Accept',
            'Access-Control-Allow-Methods': 'OPTIONS, POST, GET',
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Max-Age': 2592000,
            // ...
        };
        if (req.method === 'OPTIONS') {
            res.writeHead(204, headers);
            res.end();
        }
        else if (req.method === 'POST') {
            const matchedParams = req.url.match(controller.allowedRoomNameChars);
            const matchmakeIndex = matchedParams.indexOf(controller.matchmakeRoute);
            const method = matchedParams[matchmakeIndex + 1];
            const name = matchedParams[matchmakeIndex + 2] || '';
            const data = [];
            req.on('data', (chunk) => data.push(chunk));
            req.on('end', async () => {
                headers['Content-Type'] = 'application/json';
                res.writeHead(200, headers);
                const clientOptions = JSON.parse(Buffer.concat(data).toString());
                try {
                    const response = await controller.invokeMethod(method, name, clientOptions);
                    res.write(JSON.stringify(response));
                }
                catch (e) {
                    res.write(JSON.stringify({ code: e.code, error: e.message, }));
                }
                res.end();
            });
        }
        else if (req.method === 'GET') {
            const matchedParams = req.url.match(controller.allowedRoomNameChars);
            const roomName = matchedParams.length > 1 ? matchedParams[matchedParams.length - 1] : "";
            headers['Content-Type'] = 'application/json';
            res.writeHead(200, headers);
            res.write(JSON.stringify(await controller.getAvailableRooms(roomName)));
            res.end();
        }
    }
}

exports.Server = Server;
