// @colyseus/redis-presence/build/packages/presence/redis-presence/src/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib_es6 = require('../../../../node_modules/tslib/tslib.es6.js');
var redis = require('redis');
var util = require('util');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var redis__default = /*#__PURE__*/_interopDefaultLegacy(redis);

class RedisPresence {
    constructor(opts) {
        this.subscriptions = {};
        this.handleSubscription = (channel, message) => {
            if (this.subscriptions[channel]) {
                for (let i = 0, l = this.subscriptions[channel].length; i < l; i++) {
                    this.subscriptions[channel][i](JSON.parse(message));
                }
            }
        };
        this.sub = redis__default['default'].createClient(opts);
        this.pub = redis__default['default'].createClient(opts);
        // no listener limit
        this.sub.setMaxListeners(0);
        // create promisified pub/sub methods.
        this.subscribeAsync = util.promisify(this.sub.subscribe).bind(this.sub);
        this.unsubscribeAsync = util.promisify(this.sub.unsubscribe).bind(this.sub);
        this.publishAsync = util.promisify(this.pub.publish).bind(this.pub);
        // create promisified redis methods.
        this.smembersAsync = util.promisify(this.pub.smembers).bind(this.pub);
        this.sismemberAsync = util.promisify(this.pub.sismember).bind(this.pub);
        this.hgetAsync = util.promisify(this.pub.hget).bind(this.pub);
        this.hlenAsync = util.promisify(this.pub.hlen).bind(this.pub);
        this.pubsubAsync = util.promisify(this.pub.pubsub).bind(this.pub);
        this.incrAsync = util.promisify(this.pub.incr).bind(this.pub);
        this.decrAsync = util.promisify(this.pub.decr).bind(this.pub);
    }
    subscribe(topic, callback) {
        return tslib_es6.__awaiter(this, void 0, void 0, function* () {
            if (!this.subscriptions[topic]) {
                this.subscriptions[topic] = [];
            }
            this.subscriptions[topic].push(callback);
            if (this.sub.listeners('message').length === 0) {
                this.sub.addListener('message', this.handleSubscription);
            }
            yield this.subscribeAsync(topic);
            return this;
        });
    }
    unsubscribe(topic, callback) {
        return tslib_es6.__awaiter(this, void 0, void 0, function* () {
            const topicCallbacks = this.subscriptions[topic];
            if (!topicCallbacks) {
                return;
            }
            if (callback) {
                const index = topicCallbacks.indexOf(callback);
                topicCallbacks.splice(index, 1);
            }
            else {
                this.subscriptions[topic] = [];
            }
            if (this.subscriptions[topic].length === 0) {
                delete this.subscriptions[topic];
                yield this.unsubscribeAsync(topic);
            }
            return this;
        });
    }
    publish(topic, data) {
        return tslib_es6.__awaiter(this, void 0, void 0, function* () {
            if (data === undefined) {
                data = false;
            }
            yield this.publishAsync(topic, JSON.stringify(data));
        });
    }
    exists(roomId) {
        return tslib_es6.__awaiter(this, void 0, void 0, function* () {
            return (yield this.pubsubAsync('channels', roomId)).length > 0;
        });
    }
    setex(key, value, seconds) {
        return tslib_es6.__awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve) => this.pub.setex(key, seconds, value, resolve));
        });
    }
    get(key) {
        return tslib_es6.__awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, reject) => {
                this.pub.get(key, (err, data) => {
                    if (err) {
                        return reject(err);
                    }
                    resolve(data);
                });
            });
        });
    }
    del(roomId) {
        return tslib_es6.__awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve) => {
                this.pub.del(roomId, resolve);
            });
        });
    }
    sadd(key, value) {
        return tslib_es6.__awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve) => {
                this.pub.sadd(key, value, resolve);
            });
        });
    }
    smembers(key) {
        return tslib_es6.__awaiter(this, void 0, void 0, function* () {
            return yield this.smembersAsync(key);
        });
    }
    sismember(key, field) {
        return tslib_es6.__awaiter(this, void 0, void 0, function* () {
            return yield this.sismemberAsync(key, field);
        });
    }
    srem(key, value) {
        return tslib_es6.__awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve) => {
                this.pub.srem(key, value, resolve);
            });
        });
    }
    scard(key) {
        return tslib_es6.__awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, reject) => {
                this.pub.scard(key, (err, data) => {
                    if (err) {
                        return reject(err);
                    }
                    resolve(data);
                });
            });
        });
    }
    sinter(...keys) {
        return tslib_es6.__awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, reject) => {
                this.pub.sinter(...keys, (err, data) => {
                    if (err) {
                        return reject(err);
                    }
                    resolve(data);
                });
            });
        });
    }
    hset(key, field, value) {
        return tslib_es6.__awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve) => {
                this.pub.hset(key, field, value, resolve);
            });
        });
    }
    hincrby(key, field, value) {
        return tslib_es6.__awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve) => {
                this.pub.hincrby(key, field, value, resolve);
            });
        });
    }
    hget(key, field) {
        return tslib_es6.__awaiter(this, void 0, void 0, function* () {
            return yield this.hgetAsync(key, field);
        });
    }
    hgetall(key) {
        return tslib_es6.__awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, reject) => {
                this.pub.hgetall(key, (err, values) => {
                    if (err) {
                        return reject(err);
                    }
                    resolve(values);
                });
            });
        });
    }
    hdel(key, field) {
        return tslib_es6.__awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, reject) => {
                this.pub.hdel(key, field, (err, ok) => {
                    if (err) {
                        return reject(err);
                    }
                    resolve(ok);
                });
            });
        });
    }
    hlen(key) {
        return tslib_es6.__awaiter(this, void 0, void 0, function* () {
            return yield this.hlenAsync(key);
        });
    }
    incr(key) {
        return tslib_es6.__awaiter(this, void 0, void 0, function* () {
            return yield this.incrAsync(key);
        });
    }
    decr(key) {
        return tslib_es6.__awaiter(this, void 0, void 0, function* () {
            return yield this.decrAsync(key);
        });
    }
}

exports.RedisPresence = RedisPresence;
