// mongoose/lib/helpers/symbols.mjs
var arrayAtomicsSymbol = Symbol('mongoose#Array#_atomics');
var arrayParentSymbol = Symbol('mongoose#Array#_parent');
var arrayPathSymbol = Symbol('mongoose#Array#_path');
var arraySchemaSymbol = Symbol('mongoose#Array#_schema');
var documentArrayParent = Symbol('mongoose:documentArrayParent');
var documentIsSelected = Symbol('mongoose#Document#isSelected');
var documentIsModified = Symbol('mongoose#Document#isModified');
var documentModifiedPaths = Symbol('mongoose#Document#modifiedPaths');
var documentSchemaSymbol = Symbol('mongoose#Document#schema');
var getSymbol = Symbol('mongoose#Document#get');
var modelSymbol = Symbol('mongoose#Model');
var objectIdSymbol = Symbol('mongoose#ObjectId');
var populateModelSymbol = Symbol('mongoose.PopulateOptions#Model');
var schemaTypeSymbol = Symbol('mongoose#schemaType');
var sessionNewDocuments = Symbol('mongoose:ClientSession#newDocuments');
var scopeSymbol = Symbol('mongoose#Document#scope');
var validatorErrorSymbol = Symbol('mongoose:validatorError');

var symbols = {
	arrayAtomicsSymbol: arrayAtomicsSymbol,
	arrayParentSymbol: arrayParentSymbol,
	arrayPathSymbol: arrayPathSymbol,
	arraySchemaSymbol: arraySchemaSymbol,
	documentArrayParent: documentArrayParent,
	documentIsSelected: documentIsSelected,
	documentIsModified: documentIsModified,
	documentModifiedPaths: documentModifiedPaths,
	documentSchemaSymbol: documentSchemaSymbol,
	getSymbol: getSymbol,
	modelSymbol: modelSymbol,
	objectIdSymbol: objectIdSymbol,
	populateModelSymbol: populateModelSymbol,
	schemaTypeSymbol: schemaTypeSymbol,
	sessionNewDocuments: sessionNewDocuments,
	scopeSymbol: scopeSymbol,
	validatorErrorSymbol: validatorErrorSymbol
};

export default symbols;
export { arrayAtomicsSymbol, arrayParentSymbol, arrayPathSymbol, arraySchemaSymbol, documentArrayParent, documentIsModified, documentIsSelected, documentModifiedPaths, documentSchemaSymbol, getSymbol, modelSymbol, objectIdSymbol, populateModelSymbol, schemaTypeSymbol, scopeSymbol, sessionNewDocuments, validatorErrorSymbol };
